File: smarttrade_app/lib/features/chart_analysis/data/repositories/chart_annotation_repository_impl.dart
import 'package:dartz/dartz.dart';
import 'package:smarttrade_app/core/error/failures.dart';
import 'package:smarttrade_app/features/chart_analysis/data/datasources/chart_annotation_local_datasource.dart';
import 'package:smarttrade_app/features/chart_analysis/domain/entities/chart_annotation.dart';
import 'package:smarttrade_app/features/chart_analysis/domain/repositories/chart_annotation_repository.dart';

class ChartAnnotationRepositoryImpl implements ChartAnnotationRepository {
  const ChartAnnotationRepositoryImpl(this._localDataSource);

  final ChartAnnotationLocalDataSource _localDataSource;

  @override
  Future<Either<Failure, List<ChartAnnotation>>> getAnnotations(
    String instrumentSymbol,
  ) async {
    try {
      final annotations =
          await _localDataSource.getAnnotations(instrumentSymbol);
      return Right(annotations);
    } catch (error) {
      return Left(DatabaseFailure('Falha ao carregar marcações: $error'));
    }
  }

  @override
  Future<Either<Failure, Unit>> saveAnnotation(
    ChartAnnotation annotation,
  ) async {
    try {
      await _localDataSource.saveAnnotation(annotation);
      return const Right(unit);
    } catch (error) {
      return Left(DatabaseFailure('Falha ao salvar marcação: $error'));
    }
  }

  @override
  Future<Either<Failure, Unit>> deleteAnnotation(String id) async {
    try {
      await _localDataSource.deleteAnnotation(id);
      return const Right(unit);
    } catch (error) {
      return Left(DatabaseFailure('Falha ao remover marcação: $error'));
    }
  }
}

File: smarttrade-web/src/features/chart-analysis/data/repositories/chartAnnotationRepositoryImpl.ts
import { DatabaseFailure } from '@/core/error/failure'
import { err, ok, type Result } from '@/core/result/result'
import type { ChartAnnotation } from '../../domain/entities/chartAnnotation'
import type { ChartAnnotationRepository } from '../../domain/repositories/chartAnnotationRepository'
import { chartAnnotationLocalDataSource } from '../datasources/chartAnnotationLocalDataSource'

export class ChartAnnotationRepositoryImpl implements ChartAnnotationRepository {
  async getAnnotations(instrumentSymbol: string): Promise<Result<ChartAnnotation[]>> {
    try {
      return ok(await chartAnnotationLocalDataSource.getAnnotations(instrumentSymbol))
    } catch (error) {
      return err(new DatabaseFailure(`Falha ao carregar marcações: ${error}`))
    }
  }

  async saveAnnotation(annotation: ChartAnnotation): Promise<Result<void>> {
    try {
      await chartAnnotationLocalDataSource.saveAnnotation(annotation)
      return ok(undefined)
    } catch (error) {
      return err(new DatabaseFailure(`Falha ao salvar marcação: ${error}`))
    }
  }

  async deleteAnnotation(id: string): Promise<Result<void>> {
    try {
      await chartAnnotationLocalDataSource.deleteAnnotation(id)
      return ok(undefined)
    } catch (error) {
      return err(new DatabaseFailure(`Falha ao remover marcação: ${error}`))
    }
  }
}

export const chartAnnotationRepository: ChartAnnotationRepository =
  new ChartAnnotationRepositoryImpl()
